lib/streamlit/elements/vega_charts.py

```diff
@@ -185,12 +204,70 @@ def id_transform(data) -> dict[str, str]:
             chart_dict = altair_chart.to_dict()
 
-    # Put datasets back into the chart dict but note how they weren't
-    # transformed.
+    # Put datasets back into the chart dict:
     chart_dict["datasets"] = datasets
     return chart_dict
 
 
+def _reset_counter_pattern(prefix: str, vega_spec: str) -> str:
+    """Altair uses a global counter for unnamed parameters and views.
+    We need to reset these counters on a spec-level to make the
+    spec stable across reruns and avoid changes to the element ID.
+    """
+    pattern = re.compile(rf'"{prefix}\d+"')
+    # Get all matches without duplicates in order of appearance.
+    # Using a set here would not guarantee the order of appearance,
+    # which might lead to different replacements on each run.
+    if matches := list(dict.fromkeys(pattern.findall(vega_spec))):
+        # Add a prefix to the replacement to avoid
+        # replacing instances that already have been replaced before.
+        replacement_prefix = "__replace_prefix_o9hd101n22e1__"
+
+        # Replace all matches with a counter starting from 1
+        # We start from 1 to imitate the altair behavior.
+        for counter, match in enumerate(matches, start=1):
+            vega_spec = vega_spec.replace(
+                match, f'"{replacement_prefix}{prefix}{counter}"'
+            )
+
+        # Remove the prefix again from all replacements:
+        vega_spec = vega_spec.replace(replacement_prefix, "")
+    return vega_spec
+
+
+def _stabilize_vega_json_spec(vega_spec: str) -> str:
+    """Makes the chart spec stay stable across reruns.
+
+    Altair auto creates names for unnamed parameters & views. It uses a global counter
+    for the naming which will result in a different spec on every rerun.
+    In Streamlit, we need the spec to be stable across reruns to prevent the chart
+    from getting a new identity. So we need to replace the names with counter with a stable name.
+
+    Parameter counter:
+    https://github.com/vega/altair/blob/f345cd9368ae2bbc98628e9245c93fa9fb582621/altair/vegalite/v5/api.py#L196
+
+    View counter:
+    https://github.com/vega/altair/blob/f345cd9368ae2bbc98628e9245c93fa9fb582621/altair/vegalite/v5/api.py#L2885
+
+    This is temporary solution waiting for a fix for this issue:
+    https://github.com/vega/altair/issues/3416
+    """
+    # We only need to apply the param_ fix if there are actually parameters defined
+    # somewhere in the spec. We can check for this by looking for the '"params"' key.
+    # This isn't a perfect check, but good enough to prevent unnecessary executions
+    # for the majority of charts.
+    if '"params"' in vega_spec:
+        vega_spec = _reset_counter_pattern("param_", vega_spec)
+
+    # Simple check if the spec contains a composite chart:
+    # https://vega.github.io/vega-lite/docs/composition.html
+    # Other charts will not contain the `view_` name,
+    # so its better to not replace this pattern.
```

Comment:
suggestion: "so its better to not replace this pattern" -> "so its better to not execute the _reset_counter_pattern function unnecessarily"

Reply:
hmm, the main goal here is less on performance and more on only replacing these patterns if it is really necessary. The main reason is that there is a risk that we replace some names that are actually chosen by the user and not generated by Altair.

Reply:
I will try to clarify this in the comments

Reply:
Added more comments on top of the function